commit e5e8c15a3455c81cb16268fbaafeef3f26d2b195
Author:     Yuhong Wen <yuhongw@nvidia.com>
AuthorDate: Fri Sep 9 17:47:52 2022 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Fri Sep 9 14:47:52 2022 -0700

    Dev authz integrate (#856)
    
    * Update workspace related documentation (#684)
    
    * Update workspace related documentation
    
    * Add more details to server/client workspace and add reference
    
    * Update documentation format (#685)
    
    * add function to clean and install pip deps (#600)
    
    * add function to clean and install pip deps
    
    Co-authored-by: YanxuanLiu <yanxuanl@nvidia.com>
    
    * Update main branch CI scripts (#711)
    
    * Update main branch CI scripts
    
    * update runtest.sh
    
    * added secure logging
    
    * dev
    
    * compute complete job deploy detail
    
    * compute complete job deploy detail
    
    * consolidate security init
    
    * improve code structure
    
    * add AuditService to sub processes
    
    * changed to use FSM for session monitor
    
    * handle corner cases
    
    * support site config
    
    * use workspace object to manage file names
    
    * support site privacy
    
    * support site policy on server side
    
    * dxo filters; add show_scopes command
    
    * audit enhancement; improve dxo_filter
    
    * clean up audit; add default authz
    
    * WIP: changes for the authz.
    
    * add privacy sample
    
    * adjust fildters
    
    * Fixed broken pieces.
    
    * code style fixes.
    
    * fix typo
    
    * fix unused imports and f str
    
    * Fixed security_close() position.
    
    * corrected the name and org extraction.
    
    * Fixed Simulator with create_job_processing_context_properties.
    
    * Added the job_id to props.
    
    * Added the missing PrivacyService.
    
    * enable site custom folder.
    
    * renamed the nvflare/lighter.nvflare.py.
    
    * Fixed a merge error.
    
    * Fixed a merge error.
    
    * Removed no use import.
    
    * Fixed a duplicate name.
    
    * Fixed a sh script merge error.
    
    * Fixed unit tests.
    
    * deleted no need folders.
    
    * Removed the admin_auth_test.py.
    
    * Addressed a few comments.
    
    Co-authored-by: Yuan-Ting Hsieh (謝沅廷) <yuantinghsieh@gmail.com>
    Co-authored-by: YanxuanLiu <104543031+YanxuanLiu@users.noreply.github.com>
    Co-authored-by: YanxuanLiu <yanxuanl@nvidia.com>
    Co-authored-by: Yan Cheng <yanc@nvidia.com>
    Co-authored-by: Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
---
 nvflare/private/fed/client/client_engine.py |  27 +++---
 nvflare/private/fed/server/job_runner.py    | 145 ++++++++++++++++++++++------
 nvflare/private/fed/server/server_engine.py |  11 ++-
 3 files changed, 136 insertions(+), 47 deletions(-)

diff --git a/nvflare/private/fed/client/client_engine.py b/nvflare/private/fed/client/client_engine.py
index e7853078..f6a41c07 100644
--- a/nvflare/private/fed/client/client_engine.py
+++ b/nvflare/private/fed/client/client_engine.py
@@ -27,10 +27,12 @@ from nvflare.apis.fl_constant import MachineStatus, WorkspaceConstants
 from nvflare.apis.fl_context import FLContext, FLContextManager
 from nvflare.apis.shareable import Shareable
 from nvflare.apis.utils.common_utils import get_open_ports
+from nvflare.apis.workspace import Workspace
 from nvflare.private.admin_defs import Message
 from nvflare.private.defs import ERROR_MSG_PREFIX, ClientStatusKey, EngineConstant
 from nvflare.private.event import fire_event
-from nvflare.private.fed.utils.fed_utils import deploy_app
+from nvflare.private.fed.utils.app_deployer import AppDeployer
+from nvflare.private.fed.utils.fed_utils import security_close
 
 from .client_engine_internal_spec import ClientEngineInternalSpec
 from .client_executor import ProcessExecutor
@@ -139,15 +141,13 @@ class ClientEngine(ClientEngineInternalSpec):
         if not os.path.exists(app_root):
             return f"{ERROR_MSG_PREFIX}: Client app does not exist. Please deploy it before starting client."
 
-        if self.client.enable_byoc:
-            app_custom_folder = os.path.join(app_root, "custom")
+        app_custom_folder = os.path.join(app_root, "custom")
+        if os.path.isdir(app_custom_folder):
             try:
                 sys.path.index(app_custom_folder)
             except ValueError:
                 _remove_custom_path()
                 sys.path.append(app_custom_folder)
-        else:
-            app_custom_folder = ""
 
         self.logger.info("Starting client app. rank: {}".format(self.rank))
 
@@ -233,13 +233,17 @@ class ClientEngine(ClientEngineInternalSpec):
         self.executor.shutdown()
         return "Restart the client..."
 
-    def deploy_app(self, app_name: str, job_id: str, client_name: str, app_data) -> str:
-        workspace = os.path.join(self.args.workspace, WorkspaceConstants.WORKSPACE_PREFIX + str(job_id))
+    def deploy_app(self, app_name: str, job_id: str, job_meta: dict, client_name: str, app_data) -> str:
 
-        if deploy_app(app_name, client_name, workspace, app_data):
-            return f"Deployed app {app_name} to {client_name}"
-        else:
-            return f"{ERROR_MSG_PREFIX}: Failed to deploy_app"
+        workspace = Workspace(root_dir=self.args.workspace, site_name=client_name)
+        app_deployer = AppDeployer(
+            workspace=workspace, job_id=job_id, job_meta=job_meta, app_name=app_name, app_data=app_data
+        )
+        err = app_deployer.deploy()
+        if err:
+            return f"{ERROR_MSG_PREFIX}: {err}"
+
+        return ""
 
     def delete_run(self, job_id: str) -> str:
         job_id_folder = os.path.join(self.args.workspace, WorkspaceConstants.WORKSPACE_PREFIX + str(job_id))
@@ -286,6 +290,7 @@ def _shutdown_client(federated_client, admin_agent, touch_file):
             federated_client.process.terminate()
 
         admin_agent.shutdown()
+        security_close()
     except BaseException as e:
         traceback.print_exc()
         print("Failed to shutdown client: " + str(e))
diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index 59c664fb..8d1ed9a6 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -12,22 +12,24 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import os.path
+import json
 import shutil
 import threading
 import time
 from typing import List
 
+from nvflare.apis.client import Client
 from nvflare.apis.event_type import EventType
 from nvflare.apis.fl_component import FLComponent
-from nvflare.apis.fl_constant import FLContextKey, RunProcessKey, SystemComponents, WorkspaceConstants
+from nvflare.apis.fl_constant import AdminCommandNames, FLContextKey, RunProcessKey, SystemComponents
 from nvflare.apis.fl_context import FLContext
-from nvflare.apis.job_def import ALL_SITES, Job, RunStatus
+from nvflare.apis.job_def import ALL_SITES, Job, JobMetaKey, RunStatus
+from nvflare.apis.workspace import Workspace
 from nvflare.fuel.hci.zip_utils import zip_directory_to_bytes
-from nvflare.private.admin_defs import Message
+from nvflare.private.admin_defs import Message, MsgHeader, ReturnCode
 from nvflare.private.defs import RequestHeader, TrainingTopic
 from nvflare.private.fed.server.admin import check_client_replies
-from nvflare.private.fed.utils.fed_utils import deploy_app
+from nvflare.private.fed.utils.app_deployer import AppDeployer
 
 
 def _send_to_clients(admin_server, client_sites: List[str], engine, message):
@@ -57,16 +59,24 @@ class JobRunner(FLComponent):
         elif event_type in [EventType.JOB_COMPLETED, EventType.JOB_ABORTED, EventType.JOB_CANCELLED]:
             self._save_workspace(fl_ctx)
 
-    def _deploy_clients(self, app_data, app_name, job_id, client_sites: List[str], fl_ctx):
-        engine = fl_ctx.get_engine()
-        # deploy app to all the client sites
-        admin_server = engine.server.admin_server
+    def _make_deploy_message(self, job: Job, app_data, app_name):
         message = Message(topic=TrainingTopic.DEPLOY, body=app_data)
-        message.set_header(RequestHeader.JOB_ID, job_id)
+        message.set_header(RequestHeader.REQUIRE_AUTHZ, "true")
+
+        message.set_header(RequestHeader.ADMIN_COMMAND, AdminCommandNames.SUBMIT_JOB)
+        message.set_header(RequestHeader.JOB_ID, job.job_id)
         message.set_header(RequestHeader.APP_NAME, app_name)
-        self.log_debug(fl_ctx, f"Send deploy command to the clients for run: {job_id}")
-        replies = _send_to_clients(admin_server, client_sites, engine, message)
-        return replies
+
+        message.set_header(RequestHeader.SUBMITTER_NAME, job.meta.get(JobMetaKey.SUBMITTER_NAME))
+        message.set_header(RequestHeader.SUBMITTER_ORG, job.meta.get(JobMetaKey.SUBMITTER_ORG))
+        message.set_header(RequestHeader.SUBMITTER_ROLE, job.meta.get(JobMetaKey.SUBMITTER_ROLE))
+
+        message.set_header(RequestHeader.USER_NAME, job.meta.get(JobMetaKey.SUBMITTER_NAME))
+        message.set_header(RequestHeader.USER_ORG, job.meta.get(JobMetaKey.SUBMITTER_ORG))
+        message.set_header(RequestHeader.USER_ROLE, job.meta.get(JobMetaKey.SUBMITTER_ROLE))
+
+        message.set_header(RequestHeader.JOB_META, json.dumps(job.meta))
+        return message
 
     def _deploy_job(self, job: Job, sites: dict, fl_ctx: FLContext) -> str:
         """Deploy the application to the list of participants
@@ -76,19 +86,21 @@ class JobRunner(FLComponent):
             sites: participating sites
             fl_ctx: FLContext
 
-        Returns:
-            job_id
+        Returns:  job id
+
         """
         fl_ctx.remove_prop(FLContextKey.JOB_RUN_NUMBER)
+        fl_ctx.remove_prop(FLContextKey.JOB_DEPLOY_DETAIL)
         engine = fl_ctx.get_engine()
         run_number = job.job_id
-        workspace = os.path.join(self.workspace_root, WorkspaceConstants.WORKSPACE_PREFIX + run_number)
-        count = 1
-        while os.path.exists(workspace):
-            run_number = job.job_id + ":" + str(count)
-            workspace = os.path.join(self.workspace_root, WorkspaceConstants.WORKSPACE_PREFIX + run_number)
-            count += 1
         fl_ctx.set_prop(FLContextKey.JOB_RUN_NUMBER, run_number)
+        workspace = Workspace(root_dir=self.workspace_root, site_name="server")
+
+        client_deploy_requests = {}
+        client_token_to_name = {}
+        client_token_to_reply = {}
+        deploy_detail = []
+        fl_ctx.set_prop(FLContextKey.JOB_DEPLOY_DETAIL, deploy_detail)
 
         for app_name, participants in job.get_deployment().items():
             app_data = job.get_application(app_name, fl_ctx)
@@ -100,26 +112,82 @@ class JobRunner(FLComponent):
             client_sites = []
             for p in participants:
                 if p == "server":
-                    success = deploy_app(app_name=app_name, site_name="server", workspace=workspace, app_data=app_data)
+                    app_deployer = AppDeployer(
+                        app_name=app_name, workspace=workspace, job_id=job.job_id, job_meta=job.meta, app_data=app_data
+                    )
+
+                    err = app_deployer.deploy()
+                    if err:
+                        deploy_detail.append(f"server: {err}")
+                        raise RuntimeError(f"Failed to deploy app '{app_name}': {err}")
+
                     self.log_info(
                         fl_ctx, f"Application {app_name} deployed to the server for job: {run_number}", fire_event=False
                     )
-                    if not success:
-                        raise RuntimeError(f"Failed to deploy the App: {app_name} to the server")
+                    deploy_detail.append("server: OK")
                 else:
                     if p in sites:
                         client_sites.append(p)
 
             if client_sites:
-                replies = self._deploy_clients(app_data, app_name, run_number, client_sites, fl_ctx)
-                check_client_replies(replies=replies, client_sites=client_sites, command="deploy the App")
+                message = self._make_deploy_message(job, app_data, app_name)
+                clients, invalid_inputs = engine.validate_clients(client_sites)
+
+                if invalid_inputs:
+                    deploy_detail.append("invalid_clients: {}".format(",".join(invalid_inputs)))
+                    raise RuntimeError(f"invalid clients: {invalid_inputs}.")
+
+                for c in clients:
+                    assert isinstance(c, Client)
+                    client_token_to_name[c.token] = c.name
+                    client_deploy_requests[c.token] = message
+                    client_token_to_reply[c.token] = None
+
                 display_sites = ",".join(client_sites)
                 self.log_info(
                     fl_ctx,
-                    f"Application {app_name} deployed to the clients: {display_sites} for run: {run_number}",
+                    f"App {app_name} to be deployed to the clients: {display_sites} for run: {run_number}",
                     fire_event=False,
                 )
 
+        abort_job = False
+        if client_deploy_requests:
+            engine = fl_ctx.get_engine()
+            admin_server = engine.server.admin_server
+            client_token_to_reply = admin_server.send_requests_and_get_reply_dict(
+                client_deploy_requests, timeout_secs=admin_server.timeout
+            )
+
+            # check replies and see whether required clients are okay
+            failed_clients = []
+            for client_token, reply in client_token_to_reply.items():
+                client_name = client_token_to_name[client_token]
+                if reply:
+                    assert isinstance(reply, Message)
+                    rc = reply.get_header(MsgHeader.RETURN_CODE, ReturnCode.OK)
+                    if rc != ReturnCode.OK:
+                        failed_clients.append(client_name)
+                        deploy_detail.append(f"{client_name}: {reply.body}")
+                    else:
+                        deploy_detail.append(f"{client_name}: OK")
+                else:
+                    deploy_detail.append(f"{client_name}: unknown")
+
+            # see whether any of the failed clients are required
+            if failed_clients:
+                num_ok_sites = len(client_deploy_requests) - len(failed_clients)
+                if job.min_sites and num_ok_sites < job.min_sites:
+                    abort_job = True
+                    deploy_detail.append(f"num_ok_sites {num_ok_sites} < required_min_sites {job.min_sites}")
+                elif job.required_sites:
+                    for c in failed_clients:
+                        if c in job.required_sites:
+                            abort_job = True
+                            deploy_detail.append(f"failed to deploy to required client {c}")
+
+        if abort_job:
+            raise RuntimeError("deploy failure")
+
         self.fire_event(EventType.JOB_DEPLOYED, fl_ctx)
         return run_number
 
@@ -233,13 +301,14 @@ class JobRunner(FLComponent):
 
     def _save_workspace(self, fl_ctx: FLContext):
         job_id = fl_ctx.get_prop(FLContextKey.CURRENT_JOB_ID)
-        workspace = os.path.join(self.workspace_root, WorkspaceConstants.WORKSPACE_PREFIX + job_id)
-        workspace_data = zip_directory_to_bytes(workspace, "")
+        workspace = Workspace(root_dir=self.workspace_root)
+        run_dir = workspace.get_run_dir(job_id)
+        workspace_data = zip_directory_to_bytes(run_dir, "")
         engine = fl_ctx.get_engine()
         job_manager = engine.get_component(SystemComponents.JOB_MANAGER)
 
         job_manager.save_workspace(job_id, workspace_data, fl_ctx)
-        shutil.rmtree(workspace)
+        shutil.rmtree(run_dir)
 
     def run(self, fl_ctx: FLContext):
         engine = fl_ctx.get_engine()
@@ -262,6 +331,13 @@ class JobRunner(FLComponent):
                                 fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, ready_job.job_id)
                                 job_id = self._deploy_job(ready_job, sites, fl_ctx)
                                 job_manager.set_status(ready_job.job_id, RunStatus.DISPATCHED, fl_ctx)
+
+                                deploy_detail = fl_ctx.get_prop(FLContextKey.JOB_DEPLOY_DETAIL)
+                                if deploy_detail:
+                                    job_manager.update_meta(
+                                        ready_job.job_id, {JobMetaKey.JOB_DEPLOY_DETAIL.value: deploy_detail}, fl_ctx
+                                    )
+
                                 self._start_run(
                                     job_id=job_id,
                                     job=ready_job,
@@ -270,12 +346,19 @@ class JobRunner(FLComponent):
                                 )
                                 self.running_jobs[job_id] = ready_job
                                 job_manager.set_status(ready_job.job_id, RunStatus.RUNNING, fl_ctx)
-                            except Exception as e:
+                            except BaseException as e:
                                 if job_id:
                                     if job_id in self.running_jobs:
                                         del self.running_jobs[job_id]
                                     self._stop_run(job_id, fl_ctx)
                                 job_manager.set_status(ready_job.job_id, RunStatus.FAILED_TO_RUN, fl_ctx)
+
+                                deploy_detail = fl_ctx.get_prop(FLContextKey.JOB_DEPLOY_DETAIL)
+                                if deploy_detail:
+                                    job_manager.update_meta(
+                                        ready_job.job_id, {JobMetaKey.JOB_DEPLOY_DETAIL.value: deploy_detail}, fl_ctx
+                                    )
+
                                 self.fire_event(EventType.JOB_ABORTED, fl_ctx)
                                 self.log_error(fl_ctx, f"Failed to run the Job ({ready_job.job_id}): {e}")
 
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index b0cc1f67..e147eecb 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -54,6 +54,7 @@ from nvflare.fuel.utils import fobs
 from nvflare.private.admin_defs import Message
 from nvflare.private.defs import RequestHeader, TrainingTopic
 from nvflare.private.fed.server.server_json_config import ServerJsonConfigurator
+from nvflare.private.fed.utils.fed_utils import security_close
 from nvflare.private.scheduler_constants import ShareableHeader
 from nvflare.widgets.info_collector import InfoCollector
 from nvflare.widgets.widget import Widget, WidgetID
@@ -194,15 +195,13 @@ class ServerEngine(ServerEngineInternalSpec):
         if run_number in self.run_processes.keys():
             return f"Server run_{run_number} already started."
         else:
-            app_root = os.path.join(self._get_run_folder(run_number), self._get_server_app_folder())
+            workspace = Workspace(root_dir=self.args.workspace, site_name="server")
+            app_root = workspace.get_app_dir(run_number)
             if not os.path.exists(app_root):
                 return "Server app does not exist. Please deploy the server app before starting."
 
             self.engine_info.status = MachineStatus.STARTING
-
-            app_custom_folder = ""
-            if self.server.enable_byoc:
-                app_custom_folder = os.path.join(app_root, "custom")
+            app_custom_folder = workspace.get_app_custom_dir(run_number)
 
             open_ports = get_open_ports(2)
             self._start_runner_process(
@@ -276,6 +275,7 @@ class ServerEngine(ServerEngineInternalSpec):
         command_options = ""
         for t in args.set:
             command_options += " " + t
+
         command = (
             sys.executable
             + " -m nvflare.private.fed.app.server.runner_process -m "
@@ -784,4 +784,5 @@ def server_shutdown(server, touch_file):
         time.sleep(3.0)
     finally:
         server.status = ServerStatus.SHUTDOWN
+        security_close()
         sys.exit(2)
